afkak/common.py

```diff
@@ -41,56 +42,383 @@
 ###############
 #   Structs   #
 ###############
+
+
+@attr.frozen(hash=False, eq=False)
+class BaseStruct:
+    """
+    Base class for all structs.
+    This class behaves like a tuple to maintain backwards compatibility
+    """
+
+    def __iter__(self):
+        return iter(attr.astuple(self, recurse=False))
+
+    def __getitem__(self, item):
+        return attr.astuple(self)[item]
+
+    def __len__(self):
+        return len(attr.astuple(self))
+
+    def __eq__(self, other):
+        if isinstance(other, self.__class__):
+            return attr.astuple(self) == attr.astuple(other)
+        return NotImplemented
+
+    def __hash__(self):
+        return hash(attr.astuple(self))
+
+    def __lt__(self, other):
+        if isinstance(other, self.__class__):
+            return attr.astuple(self) < attr.astuple(other)
+        return NotImplemented
+
+
 # SendRequest is used to encapsulate messages and keys prior to
 # creating a message set
-SendRequest = namedtuple("SendRequest", ["topic", "key", "messages", "deferred"])
+@attr.frozen
+class SendRequest(BaseStruct):
+    """
+    Request to send a message to a topic
+
+    Attributes:
+        topic (str): The topic this request is for
+        key (bytes): The key for the message
+        messages (List[bytes]): The messages to include in the request
+        deferred (Deferred): The deferred to fire when the request is complete
+    """
+
+    topic: str = attr.field()
```

Comment:
The old classes are using `attr.ib` and some of the new ones are using `attr.field` and others `attr.ib` we should pick one I think. I personally prefer `attr.ib` for aesthetics but attr.field is nice because its keyword only.